File: 0163. Missing Ranges/Solution0163.java
class Solution0163 {
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> result = new ArrayList<>();
        int prev = lower - 1;
        for (int i = 0; i <= nums.length; i++) {
            int curr = (i < nums.length) ? nums[i] : upper + 1;
            if (prev + 1 <= curr - 1) {
                result.add(formatRange(prev + 1, curr - 1));
            }
            prev = curr;
        }
        return result;
    }

    // formats range in the requested format
    private String formatRange(int lower, int upper) {
        if (lower == upper) {
            return String.valueOf(lower);
        }
        return lower + "->" + upper;
    }
}


/*0 -1 1 0 2 2 4 49 51 74  76  99
 0  [ 0,  1,  3,  50,    75] 99
 s: 
 0 -1 null
 1 0  null
 2 2  2
 4 49 4->49
 51 74  51->74
 76 99  76->99
*/

File: 0163. Missing Ranges/Solution0163.cpp
#include<iostream>
#include<vector>
#include<string>

using namespace std;

class Solution { // 16
public:
    vector<string> findMissingRanges(vector<int>& nums, int lower, int upper) {
        vector<string> res;
        int prev = lower-1, n = nums.size();
        for (int i = 0; i <= n; i++) {
            int cur = (i == n) ? upper+1 : nums[i];
            if (cur - 1 >= prev + 1) {
                res.push_back(getStr(prev+1, cur-1));
            }
            prev = cur;
        }
        return res;
    }
    
    string getStr(int left, int right) {
        if (left == right) {
            return to_string(left);
        }
        string res = to_string(left) + "->" + to_string(right);
        return res;
    }
};